src/data/array/defs.lean

```diff
@@ -0,0 +1,82 @@
+| r x y := do if h : r < n ∧ r < m then do
+                let fn : fin n := ⟨r, and.elim_left h⟩,
+                let fm : fin m := ⟨r, and.elim_right h⟩,
+                y ← y.write fm <$> f (x.read fn),
+                have wf : m - (r + 1) < m - r,
+                  from nat.lt_of_succ_le $ by rw [← nat.succ_sub h.2, nat.succ_sub_succ],
+                mmap_copy_aux (r + 1) x y
+              else return y
+using_well_founded {rel_tac := λ _ _, `[exact ⟨_, measure_wf (λ a, m - a.1)⟩]}
+
+/-- Function for monadically copying part of an array to another array while applying a function.
+If `k` is a monad, and `x` and `y` are arrays of lengths `n` and `m`,
+then `array.mmap_copy f x y` will take the first `(min n m)` coefficients of `x`,
+apply the monadic function `f` to them, and write them to the corresponding of `y`.
+The result is the modified array `y`, wrapped in the monad `k`. -/
+def mmap_copy (x : array n α) (y : array m β) (f : α → k β) : k (array m β) :=
+mmap_copy_aux f 0 x y
+
+/-- Function for copying part of an array to another array while applying a function.
+If `x` is an array of length `n`, and `y` is an array of length `m`,
+then `array.map_copy f x y` will take the first `(min n m)` coefficients of `x`, apply `f` to them,
+and write them to the corresponding of `y`.
+
+See `array.copy` for the version with `f = id`. -/
+def map_copy (x : array n α) (y : array m β) (f : α → β) : array m β :=
+@mmap_copy id _ _ _ _ _ x y f
+
+/-- Function for copying part of an array to another array.
+If `x` is an array of length `n`, and `y` is an array of length `m`,
+then `array.copy f x y` will take the first `(min n m)` coefficients of `x`,
+and write them to the corresponding of `y`.
+
+See `array.map_copy` for a version that allows one to apply a function `f : α → β`
+to the coefficients of `x` before writing to `y`. -/
+def copy (x : array n α) (y : array m α) : array m α :=
+map_copy x y id
+
+/-- Auxiliary function for copying part of a list to an array while applying a function.
+If `x` is an array of length `n`, and `l` is a list of length `k`, and `m` is an index,
+then `map_copy_from_list_aux x f l m` will take the slice `l[m .. (min n k)]`, apply `f` to it,
+and write it to the corresponding slice of `y`. -/
+def map_copy_from_list_aux (x : array n β) (f : α → β) :
+  list α → ℕ → array n β
+| [] m          := x
+| (a :: rest) m := if h : m < n then
+                     (map_copy_from_list_aux rest (m + 1)).write ⟨m, h⟩ (f a)
+                   else x
+
+/-- Function for copying part of a list to an array while applying a function.
+If `x` is an array of length `n`, and `l` is a list of length `k`, and `m` is an index,
+then `array.map_copy_from_list x f l` will take the the first `(min n k)` coefficients of `l`,
+apply `f` to them, and write them to the corresponding coefficients of `x`. -/
+def map_copy_from_list (x : array n β) (f : α → β) (l : list α) : array n β :=
+map_copy_from_list_aux x f l 0
+
```

Comment:
I don't like the special cases here.  I think there are two useful functions we should add for arrays:
 - `array.mmap`
 - `array.grow`

Both can be easily (and efficiently) implemented using `array.push_back`.